enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedSinglePrimitivePipe.scala

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime.slotted.pipes
+
+import org.neo4j.collection.primitive.Primitive
+import org.neo4j.cypher.internal.compatibility.v3_4.runtime.{Slot, SlotConfiguration}
+import org.neo4j.cypher.internal.runtime.interpreted.ExecutionContext
+import org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.Expression
+import org.neo4j.cypher.internal.runtime.interpreted.pipes.{Pipe, PipeWithSource, QueryState}
+import org.neo4j.cypher.internal.runtime.slotted.SlottedExecutionContext
+import org.neo4j.cypher.internal.runtime.slotted.helpers.SlottedPipeBuilderUtils.makeSetValueInSlotFunctionFor
+import org.neo4j.cypher.internal.util.v3_4.attribution.Id
+import org.neo4j.values.AnyValue
+
+case class DistinctSlottedSinglePrimitivePipe(source: Pipe,
+                                              slots: SlotConfiguration,
+                                              toSlot: Slot,
+                                              offset: Int,
+                                              expression: Expression)
+                                             (val id: Id = Id.INVALID_ID)
+  extends PipeWithSource(source) {
+
+  //===========================================================================
+  // Compile-time initializations
+  //===========================================================================
+  private val setInSlot: (ExecutionContext, AnyValue) => Unit = makeSetValueInSlotFunctionFor(toSlot)
+
+  expression.registerOwningPipe(this)
+
+  //===========================================================================
+  // Runtime code
+  //===========================================================================
+  protected def internalCreateResults(input: Iterator[ExecutionContext],
+                                      state: QueryState): Iterator[ExecutionContext] = {
+
+
+    new Iterator[ExecutionContext] {
```

Comment:
This is a `PrefetchingIterator`. Maybe we could create a scala version of that to inherit from, so we don't have to implement it in scala more than once? E.g.

```
abstract class PrefetchingIterator[T] extends Iterator[T] {
  private var _next: T = _
  def fetchNext(): T
  def hasNext: Boolean = _next != null
  def next(): T =
    if (isEmpty)
      Iterator.empty.next() // Throw a good exception
    else {
      val toReturn = _next
      fetchNext()
      toReturn
    }
}
```